### PR TITLE
Make blank nodes reusable across expressions

### LINKED ISSUE #34: Allow blank nodes to be reused across expressions
Currently, a blank node resulting from one expression, cannot be reused in another. Following the example of `https://github.com/solid/query-ldflex/issues/33`, the following snippets return different results:

```javascript
const alice = "https://drive.verborgh.org/public/2019/blanks.ttl#Alice";
for await (const name of solid.data[alice].friends.name)
  console.log(`${name}`)
```

```javascript
const alice = "https://drive.verborgh.org/public/2019/blanks.ttl#Alice";
for await (const friend of solid.data[alice].friends)
  console.log(`${await friend.name}`)
```

This happens because Alice's friends are identified by blank nodes, and they lose context across multiple expressions. If we retry both snippets with `Alice2`, which has IRI friends, we get the same results.

---

We could strive to reuse blank nodes across expressions, by internally skolemizing them. Here is a sketch of how that could work:
- When outputting blank nodes, Comunica assigns an internal identifier to them. For instance, `_:b1` is still output as a `BlankNode`, but has a special internal field `.skolemized` that contains `urn:skolem:1234`.
- When a SPARQL query is generated from such a skolemized blank node, the skolemized IRI is used instead of a blank node.
- When returning results, any skolemized `NamedNode` is turned into a skolemized `BlankNode.

The key is inserting skolemization and deskolemization processing in the right place, for which I need to ask @rubensworks for help.

We could simply skolemize upon parsing, and then deskolemize right before results are returned. This works in all cases, _except_ when Comunica directly operates on a store (the contents of which it did not parse, so it can contain actual blank nodes).

And alternative approach is a skolemizing store wrapper. It takes a store as an argument, and translates on the fly in its `match` etc. methods.

Perhaps both approaches can be used in conjunction: skolemization in parsers for all cases, except when a store is passed, then we wrap it.

### PR DESCRIPTION
**Pull request by RubenVerborgh**
Closes #34.

### CODE REVIEW

**Comment by RubenVerborgh** [comment]
@rubensworks Would `.skolemize` also work when LDflex queries a single source? Because I see you've introduced it in a federated context (and I can imagine there to be a shortcut when there is only one source).

Concretely:
- checkout the `reusable-blank-nodes` branch of https://github.com/solid/query-ldflex/ (#64)
- `npm ln` the following modules:
  - https://github.com/LDflex/LDflex-Comunica/commit/df43615d744fe4fbf628fcea79ea174db6975c57 (or otherwise ensure Comunica >=1.11.0)
  - https://github.com/LDflex/LDflex/commit/aea6a77e2660efcb0d9916e5fa493bf74f355419
- npm run test -- test/system/blanks-test.js

Then you will see a blank node without `.skolemized`.

**Comment by rubensworks** [comment]
Hmm, this should work even for single sources. I'll look into this asap.

**Comment by rubensworks** [comment]
Updating to Comunica 1.12.1 will fix the problem 🎉

**Comment by RubenVerborgh** [comment]
@rubensworks Thanks a lot, will check!

**Comment by RubenVerborgh** [comment]
@rubensworks Lol, I've upgraded to 1.21.1 and it just _works_, without me writing any code.
I'd at least thought I would need to expose `canonical`, so now I have to spend time figuring out _why_ it works 😂 But thanks in any case!

**Comment by RubenVerborgh** [comment]
@rubensworks I first I thought it was this one:

https://github.com/comunica/comunica/blob/47c80ef2f5c118f3926c3be860865438dd8ba445/packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts#L114-L116

Given that the passed `term` is an LDflex path which always has `skolemized` defined. However, then I realized the `in` operator is not implemented yet in LDFlex, so that test always yields `false` (thankfully).

In other words, I don't think 54128f097f72c41e75f0d2e1bd265485b80bccad should work, yet it does with 1.12.1 (didn't with 1.11.0). What do you think?

**Comment by RubenVerborgh** [comment]
@rubensworks Depending on how the pattern is used, https://github.com/comunica/comunica/blob/47c80ef2f5c118f3926c3be860865438dd8ba445/packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts#L132 might be an error.

After all, a blank node in a pattern should represent a (non-returned) variable, whereas it seems to represent a constant here.

**Comment by rubensworks** [comment]
> Lol, I've upgraded to 1.12.1 and it just works, without me writing any code.

Hmm, weird. I don't _think_ it should work like that, will look into it.

> will blank nodes with the same blank node label always match?

Any blank nodes that occur at this stage originate from sources. Any blank nodes from the query will be translated to variables before query exec.

The idea is that blank nodes will be matched by label, but _not_ if they come from different sources, then they will never match.

> Also, is it safe that you are just minting variables named vs/vp/vo/vg?

This should be safe, yes, since we're working in quad pattern context here. These variable names are just used to mark wildcards in sources, but are then used to construct quads, so the variable names lever leave this context.

**Comment by RubenVerborgh** [comment]
> The idea is that blank nodes will be matched by label, but _not_ if they come from different sources, then they will never match.

I think that should be: blank nodes are matched by canonical URI if they have one. If not, blanks in a query act like variables.

**Comment by rubensworks** [comment]
> I think that should be: blank nodes are matched by canonical URI if they have one. If not, blanks in a query act like variables.

That should be what is happening now, since query-bnodes will become variables before query exec starts.

**Comment by RubenVerborgh** [comment]
Evidence seems to point at a comparison with blank node labels being done.

**Comment by rubensworks** [comment]
I just had a look, and Comunica receives queries like these:
```sparql
SELECT ?name WHERE {
  <urn:comunica_skolem:source_0:n3-0> <http://xmlns.com/foaf/0.1/name> ?name.
}
```
Which means that the blank nodes are already skolemized somehow.

I'll look into it a bit further now.

**Comment by rubensworks** [comment]
Hmm, it looks like LDflex already contains the skolemization logic here: https://github.com/LDflex/LDflex/blob/master/src/SparqlHandler.js#L209-L217

If I disable this, the test fails, as expected.

So I guess there is no issue, and everything is working as intended?
Or am I misunderstanding the problem?

**Comment by RubenVerborgh** [comment]
My bad. I had forgotten about this piece of code, which apparently uses the same `skolemized` convention already from before it was in Comunica. It all makes sense now.

So it's actually working as intended. How exciting!

Will finish this up and release soon.
